zarr/creation.py

```diff
@@ -19,32 +22,35 @@
     normalize_storage_path,
     normalize_store_arg,
 )
+from zarr._storage.store import StorageTransformer
+from zarr.sync import Synchronizer
+from zarr.types import ZARR_VERSION, DIMENSION_SEPARATOR, MEMORY_ORDER
 from zarr.util import normalize_dimension_separator
 
 
 def create(
-    shape,
-    chunks=True,
-    dtype=None,
+    shape: Union[int, Tuple[int, ...]],
+    chunks: bool = True,
+    dtype: Optional[Union[str, npt.DTypeLike]] = None,
     compressor="default",
     fill_value: Optional[int] = 0,
-    order="C",
-    store=None,
-    synchronizer=None,
-    overwrite=False,
-    path=None,
-    chunk_store=None,
-    filters=None,
-    cache_metadata=True,
-    cache_attrs=True,
-    read_only=False,
-    object_codec=None,
-    dimension_separator=None,
-    write_empty_chunks=True,
+    order: MEMORY_ORDER = "C",
+    store: Optional[Union[str, MutableMapping]] = None,
+    synchronizer: Optional[Synchronizer] = None,
+    overwrite: bool = False,
+    path: Optional[str] = None,
+    chunk_store: Optional[MutableMapping] = None,
+    filters: Optional[Sequence[Codec]] = None,
+    cache_metadata: bool = True,
+    cache_attrs: bool = True,
+    read_only: bool = False,
+    object_codec: Optional[Codec] = None,
+    dimension_separator: Optional[DIMENSION_SEPARATOR] = None,
+    write_empty_chunks: bool = True,
     *,
-    zarr_version=None,
-    meta_array=None,
-    storage_transformers=(),
+    zarr_version: Optional[ZARR_VERSION] = None,
+    meta_array: Optional[npt.ArrayLike] = None,
```

Comment:
I'm not sure that `meta_array` is an instance of `npt.ArrayLike` -- if I understand numpy's type hints correctly (and I may not!), `npt.ArrayLike` is basically `Union[Numpy array, thing that can be coerced into a numpy array]`, so I think `List[int]` counts as `ArrayLike`, but I'm not sure that `List[int]` would work for `meta_array`. cc @madsbk , who knows more about this kwarg

Reply:
Agree, `meta_array` relies on [NEP18](https://numpy.org/neps/nep-0018-array-function-protocol.html). 
Maybe a more precise hint would be something like:
```python

from numcodecs.compat import NDArrayLike
class MetaArray(NDArrayLike):
    def __array_function__(self, func, types, args, kwargs) -> Any:
        ...   
```

Reply:
👍  - I went for adding a protocol that specifies `__array_funciton__` as a method.